### ISSUE TITLE
Exclusive upper bound on the uniform distribution breaks transformed distributions

### ISSUE THREAD

**Issue by ppham27** [tensorflow/probability#219]
Consider the Weibull distribution, which we can obtain by applying the inverse of the Weibull bijector:

```python
import tensorflow as tf
import tensorflow_probability as tfp

graph = tf.Graph()
with graph.as_default():
    uniform_dist = tfp.distributions.Uniform()
    weibull_bijector = tfp.bijectors.Invert(tfp.bijectors.Weibull(
        scale=1.8,
        concentration=8.))
    weibull_dist = tfp.distributions.TransformedDistribution(
        distribution=uniform_dist, bijector=weibull_bijector)
    inverse_log_det_jacobian = weibull_bijector.inverse_log_det_jacobian([2.5, 2.6],
                                                                         event_ndims=0)
    log_prob = weibull_dist.log_prob([2.5, 2.6])
graph.finalize()

with graph.as_default(), tf.Session() as sess:
    print('Expected:', sess.run(inverse_log_det_jacobian))
    print('Actual:', sess.run(log_prob))
```

produces the output

```
Expected: [-10.055323 -14.884163]
Actual: [-10.055323       -inf]
```

This occurs because `bijector.inverse(2.6) == 1.0` ==> `uniform_dist.log_prob(bijector.inverse(2.6)) == -inf` since the support of the distribution is `[0,1)`. This is somewhat nonstandard. In R, `dunif(1) == 1`, and in SciPy, `stats.uniform.logpdf(1) == 0.0`.

My preference would to be consistent with other statistical software and have the upper bound be inclusive. Failing that, we could add a default argument to `tfp.distributions.Uniform(..., upper=False)`. If that's still not desirable, the workaround isn't terrible. I could construct a uniform distribution like `tfp.distributions.Uniform(high=1. + 1e-7)`.

In an ideal world, perhaps `bijector.inverse(y) < 1.0` would always be true, but that seems unlikely in float32 precision. Happy to submit the PR myself if we can come to some agreement.

The corresponding code in SciPy of course works fine:

```python
from scipy import stats

stats.weibull_min.logpdf([2.5, 2.6], scale=1.8, c=8.)
```

outputs `array([-10.05531725, -14.88416383])`.

**Comment by brianwa84** [comment]
Do you need an inclusive lower bound too? If not, you could take 1-uniform using the Affine bijector.

**Comment by brianwa84** [comment]
How problematic would it be to have Uniform.log_prob support the point, but
Uniform.sample not support it? Seems bad.

**Comment by axch** [comment]
I'm actually not sure it would be that bad.  Having the sampler never emit 1.0 while the density function claims its density is 1 would amount to an O(1)-ulp total variation distance bug in the sampler, which is O(2^-52) in 64-bit and O(2^-23) in 32-bit precision.  My understanding is that we are comfortable with various other sampler bugs of that scale elsewhere in the codebase.

**Comment by ppham27** [comment]
If it's important that Uniform.sample can actually generate the upper bound, one option could be to make a change here:

https://github.com/tensorflow/probability/blob/ccc770782c1d08bb86e2d21c99d11d58d456e1ce/tensorflow_probability/python/distributions/uniform.py#L158-L163

You could call `tf.random_uniform(..., minval=self.low, maxval=self.high + 2.**-23)`. In float64, it's `2**-52`. In the more general case when `low` and `high` are not 0 and 1, adding a specific smaller number isn't quite right. You'd probably actually need use bit operations to flip a bit in the significand.

I don't really like this idea, but just a suggestion.

**Comment by axch** [comment]
Log-prob function fixed by 553b13194b11551d5ae5bbd53fb441c3d91b0991.
